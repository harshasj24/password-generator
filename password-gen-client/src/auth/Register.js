import Visibility from "@mui/icons-material/Visibility";
import VisibilityOff from "@mui/icons-material/VisibilityOff";
import {
  Button,
  FormControl,
  FormControlLabel,
  IconButton,
  InputAdornment,
  InputLabel,
  Modal,
  OutlinedInput,
  TextField,
} from "@mui/material";
import { ErrorMessage, Form, Formik } from "formik";
import React, { useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import VerifiedIcon from "@mui/icons-material/Verified";
import NewReleasesIcon from "@mui/icons-material/NewReleases";

import * as yup from "yup";
import axios from "axios";

const Register = () => {
  const navigate = useNavigate();
  const [isError, setIsError] = useState(false);
  const [values, setValues] = useState({
    password: "",
    showPassword: false,
  });
  const [entredPass, setEntredPass] = useState();
  const [open, setOpen] = useState(false);
  const initialVlaues = {
    fName: "",
    lName: "",
    email: "",
    password: "",
    cPassword: "",
  };
  const handleChange = (prop, setFieldValue) => (event) => {
    setValues({ ...values, [prop]: event.target.value });
    setFieldValue("password", event.target.value);
    setEntredPass(event.target.value);
  };
  const handleClickShowPassword = () => {
    setValues({
      ...values,
      showPassword: !values.showPassword,
    });
  };
  const handleMouseDownPassword = (event) => {
    event.preventDefault();
  };
  const handelClick = () => {
    setOpen(false);
    navigate("/login");
  };
  const validationSchema = yup.object({
    fName: yup
      .string()
      .required("please provide your first name")
      .max(20, "First name should contain only 20 characters"),
    lName: yup
      .string()
      .required("please provide your last name")
      .max(20, "Last name should contain only 20 characters"),
    email: yup
      .string()
      .required("Please provide your email address")
      .email("Please provide a valid email address"),

    password: yup
      .string()
      .required("Please provide a password")
      .matches(
        /(?=.*[A-Z])(?=.*[a-z])(?=.*[1-9])(?=.*[!@#$%^&*])[a-zA-Z/d!@#$%^&*]/,
        "Password must contain one uppercase, one lowercase, one speical character, one number and minium of 8 charcters"
      ),
    cPassword: yup
      .string()
      .required("Please confirm your password")
      .oneOf([yup.ref("password"), null], "password must match"),
  });

  const handelSubmit = async (values) => {
    try {
      await axios.post(" http://localhost:4500/users/signUp", values);
      setIsError(false);
      setOpen(true);
    } catch (error) {
      setIsError(true);
      setOpen(true);
    }
  };
  return (
    <div>
      <h3 className="mb-3">Register</h3>
      <Formik
        initialValues={initialVlaues}
        validationSchema={validationSchema}
        onSubmit={handelSubmit}
      >
        {({ setFieldValue }) => {
          return (
            <Form data-testid="regForm">
              <TextField
                name={"fName"}
                size="small"
                label="First Name "
                variant="outlined"
                className="w-100 "
                onChange={(e) => {
                  setFieldValue("fName", e.target.value);
                }}
              />
              <ErrorMessage name="fName">
                {(msg) => <span className="text-danger">*{msg}</span>}
              </ErrorMessage>
              <TextField
                size="small"
                name={"lName"}
                label="Last Name"
                variant="outlined"
                className="w-100 my-3"
                onChange={(e) => {
                  setFieldValue("lName", e.target.value);
                }}
              />
              <ErrorMessage name="lName">
                {(msg) => <span className="text-danger">*{msg}</span>}
              </ErrorMessage>
              <TextField
                size="small"
                name={"email"}
                label="Email"
                variant="outlined"
                className="w-100 "
                onChange={(e) => {
                  setFieldValue("email", e.target.value);
                }}
              />
              <ErrorMessage name="email">
                {(msg) => <span className="text-danger">*{msg}</span>}
              </ErrorMessage>
              <FormControl
                size="small"
                className="my-3"
                sx={{ width: "100%" }}
                variant="outlined"
              >
                <InputLabel htmlFor="outlined-adornment-password">
                  Password
                </InputLabel>
                <OutlinedInput
                  id="outlined-adornment-password"
                  type={values.showPassword ? "text" : "password"}
                  value={values.password}
                  onChange={handleChange("password", setFieldValue)}
                  endAdornment={
                    <InputAdornment position="end">
                      <IconButton
                        aria-label="toggle password visibility"
                        onClick={handleClickShowPassword}
                        onMouseDown={handleMouseDownPassword}
                        edge="end"
                      >
                        {values.showPassword ? (
                          <VisibilityOff />
                        ) : (
                          <Visibility />
                        )}
                      </IconButton>
                    </InputAdornment>
                  }
                  label="Password"
                />
                <ErrorMessage name="password">
                  {(msg) => <span className="text-danger">*{msg}</span>}
                </ErrorMessage>
              </FormControl>
              <TextField
                size="small"
                name={"confirm password"}
                label="confirmPassword"
                variant="outlined"
                className="w-100"
                onChange={(e) => {
                  setFieldValue("cPassword", e.target.value);
                }}
              />
              <ErrorMessage name="cPassword">
                {(msg) => <span className="text-danger">*{msg}</span>}
              </ErrorMessage>
              <Button
                variant="contained"
                className="d-flex mt-3 ml-auto"
                color="success"
                type="submit"
                data-testid="formSubmit"
              >
                Sign Up
              </Button>
            </Form>
          );
        }}
      </Formik>

      <p className="text-center">
        Have an account?{" "}
        <Link to={"/login"}>
          {" "}
          <strong>LOGIN</strong>{" "}
        </Link>
      </p>

      <div className="modal ">
        <Modal open={open}>
          <div className="reg bg-light w-25 mx-auto  h-50 mt-5 p-3 justify-content-center">
            {!isError ? (
              <div className="reg__success d-flex flex-column align-items-center justify-content-center h-100">
                <VerifiedIcon
                  className="success-icon"
                  color="success"
                  sx={{ fontSize: "120px" }}
                />
                <h4 className="my-3">Registration Successfull!</h4>
                <Button
                  onClick={handelClick}
                  variant="contained"
                  color="success"
                >
                  Login
                </Button>
              </div>
            ) : (
              <div className="ref-error d-flex flex-column align-items-center justify-content-center h-100 ">
                <NewReleasesIcon color="error" sx={{ fontSize: "120px" }} />
                <h4>Something went wrong</h4>

                <p style={{ fontSize: "15px" }}>
                  Email you have enter is already regestred{" "}
                </p>
                <Link
                  onClick={() => {
                    setOpen(false);
                  }}
                  to={"/signUp"}
                >
                  Try again
                </Link>
              </div>
            )}
          </div>
        </Modal>
      </div>
    </div>
  );
};

export default Register;
