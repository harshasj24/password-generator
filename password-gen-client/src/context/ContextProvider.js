import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";

export const passwordCotext = createContext();

export const ContextProvider = ({ children }) => {
  const [sliderVal, setSliderVal] = useState(12);
  const [password, setPassword] = useState("");
  const [breakPoint, setBreakPoint] = useState(null);
  const [isLoading, setisloading] = useState(false);
  const breakPoints = {
    mobile: "(max-width:600px)",
    tablet: "(min-width:600px) and (max-width:960px)",
    laptop: "(min-width:960px) and (max-width:1023px)",
    desktop: "(min-width:1024px)",
  };

  const copyToClipBoard = (text) => {
    window.navigator.clipboard.writeText(text);
  };

  const matchMediaQuerry = (breakPoints, setBreakPoint) => {
    Object.keys(breakPoints).map((key) => {
      if (window.matchMedia(breakPoints[key]).matches) {
        setBreakPoint(key);
      }
    });
  };

  const breakPointObserver = () => {
    window.addEventListener("resize", () => {
      matchMediaQuerry(breakPoints, setBreakPoint);
    });
  };

  const values = useMemo(() => {
    return {
      password,
      setPassword,
      sliderVal,
      setSliderVal,
      copyToClipBoard,
      breakPointObserver,
      breakPoint,
      isLoading,
      setisloading,
    };
  });

  return (
    <div>
      <passwordCotext.Provider value={values}>
        {children}
      </passwordCotext.Provider>
    </div>
  );
};

export const usePassword = () => {
  return useContext(passwordCotext);
};
