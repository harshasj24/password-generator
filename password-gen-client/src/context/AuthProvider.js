import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import { useNavigate } from "react-router-dom";
import { useLocalStorage } from "../hooks/UseLocalStorage";
import axios from "axios";
import { useVault } from "./VaultProvider";
const AuthContext = createContext();

export const AuthProvider = ({ children }) => {
  // const navigate = useNavigate();
  // const [isLoggedIn, setIsLoggedin] = useState(false);
  const [user, setUser] = useLocalStorage("user", null);
  useEffect(() => {
    console.log(user);
  }, [user]);
  const login = async (userDetails) => {
    setUser(userDetails);
    console.log(userDetails, "auth-provider");
  };
  const logout = async () => {
    setUser(null);
    // console.log(responce);
  };
  const value = useMemo(() => {
    return { user, login, logout, setUser };
  }, [user, login]);
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export const useAuth = () => {
  return useContext(AuthContext);
};
