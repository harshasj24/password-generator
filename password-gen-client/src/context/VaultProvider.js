import axios from "axios";
import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import { useNavigate } from "react-router-dom";
import { useAuth } from "../context/AuthProvider";
import environemnt from "../environments/environment";
const vaultContex = createContext();
export const VaultProvider = ({ children }) => {
  const [passwords, setPasswords] = useState([]);
  const { user } = useAuth();
  const navigate = useNavigate();
  const baseUrl = environemnt.apiUrl;
  const [updatePass, setUpdatePass] = useState({
    action: null,
    isUpdating: false,
    _id: null,
    pName: "",
  });
  const [snackbar, setSnackbar] = useState({
    open: false,
    message: "",
  });
  const [modalDetails, setModalDetails] = useState({
    open: false,
    _id: null,
    type: "",
  });

  const handleModalDetails = (details) => {
    setModalDetails({ ...modalDetails, ...details });
  };

  const handleSnackbar = (open, message) => {
    setSnackbar({ ...snackbar, open: open, message: message });
  };
  const [slectedId, setSelectedId] = useState(null);
  useEffect(() => {
    console.log(user, "effect");
  }, [user]);

  const headers = user && {
    Authorization: "barrer " + user.token,
  };

  const savePassword = async (PasswordDetails) => {
    try {
      const repsonce = await axios.post(
        `${baseUrl}/vault/savePasswords`,
        { ...PasswordDetails, _id: user.id },
        {
          headers: headers,
        }
      );
      const newPassword = {
        ...PasswordDetails,
        _id: repsonce.data.data._id,
        date: repsonce.data.data.date,
        password: repsonce.data.data.cipherText,
        status: repsonce.data.data.status,
      };

      setPasswords([...passwords, newPassword]);
    } catch (error) {
      console.log(error);
    }
  };
  const isUpdating = (details) => {
    setUpdatePass({
      ...updatePass,
      ...details,
    });
  };
  const resetUpdating = () => {
    setUpdatePass({
      ...updatePass,
      isUpdating: false,
      _id: null,
      pName: "",
    });
  };
  const updatePassword = async (passwordDetails) => {
    const { password, pName } = passwordDetails;
    const reqData = {
      password,
      _id: updatePass._id,
      pName,
    };
    const responece = await axios.put(
      `${baseUrl}/vault/updatePassword`,
      reqData,
      {
        headers: headers,
      }
    );

    const index = passwords.findIndex((data) => {
      return data._id === updatePass._id;
    });
    console.log(index);

    const newPasswords = [...passwords];
    newPasswords[index] = {
      ...newPasswords[index],
      password: responece.data.data.cipherText,
      date: responece.data.data.date,
      status: responece.data.data.status,
      pName,
    };
    setPasswords([...newPasswords]);
    handleSnackbar(true, "Password Updated Sucessfully");
  };

  const getAllPasswords = async () => {
    try {
      if (passwords.length === 0) {
        const repsonce = await axios.get(
          `${baseUrl}/vault/password/${user?.id}`,
          {
            headers: headers,
          }
        );
        console.log(repsonce);

        setPasswords(repsonce.data.data);
      }
    } catch (error) {
      console.log(error);
      setPasswords([]);
    }
  };
  // validatePassword
  const checkPassword = async (data) => {
    try {
      const responce = await axios.post(
        `${baseUrl}/users/validatePassword`,
        data,
        {
          headers: headers,
        }
      );
      return responce;
    } catch (error) {
      if (error.status === 403) {
        navigate("/access");
      }
      return false;
    }
  };

  const deletePassword = async () => {
    try {
      const responce = await axios.delete(
        `${baseUrl}/vault/delete/${updatePass._id}`,
        {
          headers: headers,
        }
      );
      const newPasswords = [...passwords].filter(
        (val) => val._id !== updatePass._id
      );
      setPasswords(newPasswords);
      handleSnackbar(true, "Password Deleted Sucessfully");
    } catch (error) {
      console.log(error);
    }
  };

  const loopElements = (range, element) => {
    for (let i = 0; i < range; i++) {
      console.log(i, range, "vault");
      return element;
    }
  };

  const getOnePassword = async (_id) => {
    const responce = await axios.get(`${baseUrl}/vault/getPassword/${_id}`, {
      headers: headers,
    });
    return responce;
  };
  const values = useMemo(() => {
    return {
      passwords,
      setPasswords,
      savePassword,
      getAllPasswords,
      updatePass,
      isUpdating,
      updatePassword,
      headers,
      resetUpdating,
      checkPassword,
      slectedId,
      setSelectedId,
      deletePassword,
      pName: updatePass.pName,
      loopElements,
      getOnePassword,
      snackbar,
      handleSnackbar,
      modalDetails,
      handleModalDetails,
    };
  }, [
    passwords,
    updatePass,
    user,
    slectedId,
    snackbar.open,
    modalDetails,
    handleSnackbar,
  ]);

  return <vaultContex.Provider value={values}>{children}</vaultContex.Provider>;
};

export const useVault = () => {
  return useContext(vaultContex);
};
