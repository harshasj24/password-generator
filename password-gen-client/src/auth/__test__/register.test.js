import { fireEvent, render, screen } from "@testing-library/react";
import axios from "axios";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import { AuthProvider, useAuth } from "../../context/AuthProvider";
import { ContextProvider } from "../../context/ContextProvider";
import { VaultProvider } from "../../context/VaultProvider";
import Register from "../Register";
jest.mock("axios", () => {
  return { ...jest.requireActual("axios"), post: jest.fn() };
});

let component;
beforeEach(() => {
  component = (
    <BrowserRouter>
      <AuthProvider>
        <VaultProvider>
          <ContextProvider>
            <Register />
          </ContextProvider>
        </VaultProvider>
      </AuthProvider>
    </BrowserRouter>
  );
});

describe("Register Component", () => {
  it("it should render the component", () => {
    const tree = renderer.create(component).toJSON;
    expect(tree).toMatchSnapshot();
  });
  it("it should render the register form", () => {
    const wrapper = render(component);
    const registerFrom = screen.getByTestId("regForm");
    expect(registerFrom).toBeInTheDocument();
  });

  it("it should submit the form", () => {
    const componentLo = render(component);
    const fn = jest.fn();
    let btn = screen.getByTestId("formSubmit");
    fireEvent.click(btn, { Event: fn() });
    expect(btn).toBeTruthy();
    expect(fn).toHaveBeenCalledTimes(1);
  });
});
