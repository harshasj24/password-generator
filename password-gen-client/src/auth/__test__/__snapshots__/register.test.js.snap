// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Component it should render the component 1`] = `[Function]`;
